A105-max_concurrent_streams-connection-scaling.md

```diff
@@ -0,0 +1,629 @@
+max_connections_per_subchannel attribute is larger than the channel's
+limit, it will be capped to that limit.  This capping will be performed
+in the subchannel itself, so that it will apply regardless of where the
+attribute is set.
+
+The max_connections_per_subchannel attribute can change with each resolver
+update, regardless of whether it is set via the service config or via an
+LB policy.  When this happens, we do not want to throw away the subchannel
+and create a new one, since that would cause unnecessary connection churn.
+This means that the max_connections_per_subchannel attribute must not
+be considered part of the subchannel's unique identity that is set only
+at subchannel creation time; instead, it must be an attribute that can
+be changed over the life of a subchannel.  The approach for this will be
+different in C-core than in Java and Go.
+
+If the max_connections_per_subchannel attribute is unset, the subchannel
+will assume a default of 1, which effectively means the same behavior
+as before this gRFC.
+
+#### C-core
+
+In C-core, every time there is a resolver update, the LB policy
+calls `CreateSubchannel()` for every address in the new address list.
+The `CreateSubchannel()` call returns a subchannel wrapper that holds
+a ref to the underlying subchannel.  The channel uses a subchannel
+pool to store the set of currently existing subchannels: the requested
+subchannel is created only if it doesn't already exist in the pool;
+otherwise, the returned subchannel wrapper will hold a new ref to the
+existing subchannel, so that it doesn't actually wind up creating a new
+subchannel (only a new subchannel wrapper).  This means that we do not
+want the max_connections_per_subchannel attribute to be part of the
+subchannel's key in the subchannel pool, or else we will wind up
+recreating the subchannel whenever the attribute's value changes.
+
+In addition, by default, C-core's subchannel pool is shared between
+channels, meaning that if two channels attempt to create the same
+subchannel, they will wind up sharing a single subchannel.  In this
+case, each channel using a given subchannel may have a different value
+for the max_connections_per_subchannel attribute.  The subchannel will
+use the maximum value set for this attribute across all channels.
+
+To support this, the implementation will be as follows:
+- The subchannel will store a map from max_connections_per_subchannel
+  value to the number of subchannel wrappers currently holding a ref to
+  it with that value.  Entries are added to the map whenever the first
+  ref for a given value of max_connections_per_subchannel is taken, and
+  entries are removed from the map whenever the last ref for a given
+  value of max_connections_per_subchannel is removed.  Whenever the set
+  of entries changes, the subchannel will do a pass over the map to find
+  the new max value to use.
+- We will create a new channel arg called
```

Comment:
Will this apply to `grpc::CreateCustomChannel()`? If so then how would coexist with supplying custom ServiceConfig (`SetServiceConfigJSON()`).